### PR TITLE
Fix sort on key type, closes #204

### LINKED ISSUE #204: Sort on key type fix
- Parent issues: -
- Relevant issues: -
- Estimated effort: 2 hours

* * *

## User story
As a user I want to see when a certain text bar can be unfolded to show certain elements so that I can find for example the contacts that are ordered under RSAKey or DSAKey.

## What
This issue solves the bug that now arrow-down is shown in the text bar and that the correct users are drawn under the correct text bar in MainActivity.

## Why
This issue is needed for the project because the current implementation doesn't work and is not straight-forward because the arrow is missing.

## How
This feature can be added by adding the down-arrow and by letting the correct contacts appear when clicked on an unfoldable text bar.

## Definition of ready
This features is not ready until the following issues are finished:

- 

## Definition of done
This task can be considered done when

- Contacts can be sorted on key type
- The unfoldable key bars indicate that they're unfoldable by drawing a down-arrow

### PR DESCRIPTION
**Pull request by KilianCallebaut**
- Relevant Issues: #204 
- Related Pull Requests: -

* * *

## What

This Pull Request adds to the repository the fix to sort contacts by the keytypes they posess

## Why

This Pull Request is needed because it was broken and now it's fixed.

## How

This feature can be viewed/tested within the project by trying out the sort button with contacts with different keytypes

## Alternative implementation

Other implementations that I've have considered are ...

## Notes

None

### CODE REVIEW

File: app/src/main/java/com/nervousfish/nervousfish/activities/CreateProfileActivity.java [hunk 1]
```diff
@@ -39,8 +41,13 @@
  * The {@link android.app.Activity} that is used to create a user profile when the app is first
  * used.
  */
-@SuppressWarnings("checkstyle:ReturnCount")
-//Suppresses return count to allow multiple returncodes while checking input fields.
+@SuppressWarnings({"checkstyle:ReturnCount", "PMD.ExcessiveImports"})
+/*
+1) Suppresses return count to allow multiple returncodes while checking input fields.
+2) Suppress excessive imports because it's necessairy and the 2 added imports methods would be unlogical
+    to outsource to another class
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/CreateProfileActivity.java]: Can you change these so they're in line with other comments of this kind 😇

File: app/src/main/java/com/nervousfish/nervousfish/activities/CreateProfileHelper.java [hunk 2]
```diff
@@ -40,7 +40,7 @@
     }
 
     /**
-     * Generates a KeyPair based on the type selected.
+     * Generates a list with one {@link KeyPair} based on the type selected.
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/CreateProfileHelper.java]: `* Generates a list of {@link KeyPair}s...`

Even if it is currently always one, in the future this is not necessarily true. If you do want to always give one `KeyPair`, you simply shouldn't return a `List`

In regards to the function after this, that one shouldn't be needed...

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivitySorter.java [hunk 3]
```diff
@@ -134,24 +122,17 @@ private void hideNoContactsContent() {
     private void sortOnKeyType() {
         final ExpandableListView ev = (ExpandableListView) this.mainActivity.findViewById(R.id.expandable_contact_list_by_key_type);
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivitySorter.java]: Could you change the name of this local variable to something more descriptive 😇

File: app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java [hunk 4]
```diff
@@ -48,19 +48,20 @@
      * A preference value change listener that updates the preference's summary
      * to reflect its new value.
      */
-    private static Preference.OnPreferenceChangeListener sBindPreferenceSummaryToValueListener = new Preference.OnPreferenceChangeListener() {
+    private static final Preference.OnPreferenceChangeListener S_BIND_PREFERENCE_SUMMARY_TO_VALUE_LISTENER =
+        new Preference.OnPreferenceChangeListener() {
         @Override
         public boolean onPreferenceChange(final Preference preference, final Object value) {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SettingsActivity.java]: Can you update the indentation here 😇

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapterLoader.java [hunk 5]
```diff
@@ -61,7 +61,7 @@ private String readDatabaseToString() throws IOException {
         final StringBuilder databaseFileStringBuilder = new StringBuilder();
 
         // Get the database from the database file
-        try (BufferedReader databaseReader = (BufferedReader) this.fileSystem.getReader(databasePath)) {
+        try (BufferedReader databaseReader = (BufferedReader) this.fileSystem.getReader(this.databasePath)) {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapterLoader.java]: Could you remove this enter while you're at it 😇

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapterLoader.java [hunk 6]
```diff
@@ -104,9 +104,9 @@ private boolean checkPassword(final String password) throws IOException {
      * @return The secretKey made from the password\
      */
     public SecretKey loadKey(final String password) throws IOException, InvalidKeySpecException {
-        checkPassword(password);
-        if (checkPassword(password)) {
-            return encryptor.makeKeyFromPassword(password);
+        this.checkPassword(password);
+        if (this.checkPassword(password)) {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapterLoader.java]: Two times `this.checkPassword(password)`?

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/INfcHandler.java [hunk 7]
```diff
@@ -1,14 +1,13 @@
 package com.nervousfish.nervousfish.modules.pairing;
 
-import com.nervousfish.nervousfish.modules.IModule;
-
 /**
  * Defines a module used for pairing using NFC that can be used by a service locator.
  */
-public interface INfcHandler extends IPairingHandler, IModule {
+public interface INfcHandler extends IPairingHandler {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/INfcHandler.java]: Why did you change this?

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/IPairingHandler.java [hunk 8]
```diff
@@ -8,7 +8,7 @@
 /**
  * Defines the interface implemented by all pairing handlers
  */
-interface IPairingHandler extends IModule, Serializable {
+interface IPairingHandler extends IModule {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/IPairingHandler.java]: Why did you change this?

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/IQRHandler.java [hunk 9]
```diff
@@ -1,10 +1,8 @@
 package com.nervousfish.nervousfish.modules.pairing;
 
-import com.nervousfish.nervousfish.modules.IModule;
-
 /**
  * Defines a module used for pairing using a QR code that can be used by a service locator.
  */
 
-public interface IQRHandler extends IPairingHandler, IModule {
+public interface IQRHandler extends IPairingHandler {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/IQRHandler.java]: Why did you change this?

Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/IQRHandler.java]: Because IPairingHandler extends IModule

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/IPairingHandler.java (hunk 8 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/IPairingHandler.java]: Because IModule extends Serializable

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/INfcHandler.java (hunk 7 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/INfcHandler.java]: Because IPairingHandler extends IModule

File: app/src/main/java/com/nervousfish/nervousfish/activities/CreateProfileHelper.java (hunk 2 above)
Comment (KilianCallebaut) [inline app/src/main/java/com/nervousfish/nervousfish/activities/CreateProfileHelper.java]: I added the function after that to facilitate the generating of multiple keys in create contact, is it bad that that one exists?

Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/CreateProfileHelper.java]: I understand you added it for that reason, but I don't see why we would need one method that returns a `List` with a single key and one that returns a `List` with multiple keys. Simply always return a `List` of keys, and the rest of the application should be able to handle that. imo.

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/INfcHandler.java (hunk 7 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/INfcHandler.java]: 👍

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/IPairingHandler.java (hunk 8 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/IPairingHandler.java]: 👍

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/IQRHandler.java (hunk 9 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/IQRHandler.java]: 👍

File: app/src/androidTest/assets/features/main.feature [hunk 10]
```diff
@@ -70,3 +70,8 @@ Feature: Main Activity
     Given I am viewing the main activity
     When I click the sort button in the main activity
     Then The app shouldn't crash
+
+  Scenario: Clicking the sort button with different key types
+    Given I am viewing the main activity
+    When I click the sort button in the main activity while there are contacts with different keys in the database
```
Comment (ericcornelissen) [inline app/src/androidTest/assets/features/main.feature]: You should split this up into a `when` and a `then` 😉
